domain/user/state/state.go

```diff
@@ -0,0 +1,452 @@
+// is returned that satisfies usererrors.NotFound.
+func (st *State) SetActivationKey(ctx context.Context, uuid user.UUID, activationKey string) error {
+	db, err := st.DB()
+	if err != nil {
+		return fmt.Errorf("cannot get DB access: %w", err)
+	}
+
+	return db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		removed, err := st.isRemoved(ctx, tx, uuid)
+		if err != nil {
+			return fmt.Errorf("cannot get user %q: %w", uuid, err)
+		}
+		if removed {
+			return fmt.Errorf("cannot set activation key for removed user %q", uuid)
+		}
+
+		err = removePasswordHash(ctx, tx, uuid)
+		if err != nil {
+			return errors.Trace(err)
+		}
+
+		return errors.Trace(setActivationKey(ctx, tx, uuid, activationKey))
+	})
+}
+
+// SetPasswordHash removes any active activation keys and sets the user
+// password hash and salt. If no user is found for the supplied name an error
+// is returned that satisfies usererrors.NotFound.
+func (st *State) SetPasswordHash(ctx context.Context, uuid user.UUID, passwordHash string, salt []byte) error {
+	db, err := st.DB()
+	if err != nil {
+		return fmt.Errorf("cannot get DB access: %w", err)
+	}
+
+	return db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		removed, err := st.isRemoved(ctx, tx, uuid)
+		if err != nil {
+			return fmt.Errorf("cannot get user uuid %q: %w", uuid, err)
+		}
+		if removed {
+			return fmt.Errorf("cannot set password hash for removed user %q", uuid)
+		}
+
+		err = removeActivationKey(ctx, tx, uuid)
+		if err != nil {
+			return errors.Trace(err)
+		}
+
+		return errors.Trace(setPasswordHash(ctx, tx, uuid, passwordHash, salt))
+	})
+}
+
+func addUser(ctx context.Context, tx *sqlair.TX, uuid user.UUID, user user.User, creatorUuid user.UUID) error {
+	addUserQuery := `
+INSERT INTO user (uuid, name, display_name, created_by_uuid, created_at) 
+VALUES ($M.uuid, $M.name, $M.display_name, $M.created_by_uuid, $M.created_at)
+`
+
+	insertAddUserStmt, err := sqlair.Prepare(addUserQuery, sqlair.M{})
+	if err != nil {
+		return fmt.Errorf("cannot prepare query: %w", err)
+	}
+
+	err = tx.Query(ctx, insertAddUserStmt, sqlair.M{
+		"uuid":            uuid.String(),
+		"name":            user.Name,
+		"display_name":    user.DisplayName,
+		"created_by_uuid": creatorUuid.String(),
+		"created_at":      time.Now()},
+	).Run()
+	if databaseutils.IsErrConstraintUnique(err) {
+		return usererrors.AlreadyExists
+	} else if databaseutils.IsErrConstraintForeignKey(err) {
+		return usererrors.UserCreatorUUIDNotFound
+	} else if err != nil {
+		return fmt.Errorf("cannot insert user %q: %w", user.Name, err)
+	}
+
+	return nil
+}
+
+func setPasswordHash(ctx context.Context, tx *sqlair.TX, uuid user.UUID, passwordHash string, salt []byte) error {
+	defineUserAuthenticationQuery := `
+INSERT INTO user_authentication (user_uuid, disabled) 
+VALUES ($M.uuid, $M.disabled)
+ON CONFLICT(user_uuid) DO UPDATE SET user_uuid = excluded.user_uuid, disabled = excluded.disabled
```

Comment:
If we already have an inserted user in the `user_authentication` we should do nothing.

```suggestion
ON CONFLICT(user_uuid) IGNORE
```